localstack/services/forwarder.py

```diff
@@ -0,0 +1,220 @@
+"""
+This module contains utilities to call a backend (e.g., an external service process like DynamoDBLocal)
+from
+"""
+from typing import Any, Callable, Mapping, Optional, Tuple, Union
+from urllib.parse import urlsplit
+
+import requests
+from botocore.awsrequest import AWSPreparedRequest
+from botocore.parsers import create_parser
+from werkzeug.datastructures import Headers
+
+from localstack import config
+from localstack.aws.api.core import (
+    CommonServiceException,
+    HttpRequest,
+    HttpResponse,
+    RequestContext,
+    ServiceRequest,
+    ServiceRequestHandler,
+    ServiceResponse,
+)
+from localstack.aws.skeleton import DispatchTable, create_dispatch_table
+from localstack.aws.spec import load_service
+from localstack.utils.aws import aws_stack
+from localstack.utils.strings import to_bytes, to_str
+
+HttpBackendResponse = Tuple[int, dict, Union[str, bytes]]
+ServiceRequestOrMapping = Union[Mapping[str, Any], ServiceRequest]
+RequestForwarder = Callable[[RequestContext, Optional[ServiceRequestOrMapping]], HttpResponse]
+
+
+def ForwardingFallbackDispatcher(
+    provider: object, request_forwarder: RequestForwarder
+) -> DispatchTable:
+    """
+    Wraps a provider with a request forwarder. It does by creating a new DispatchTable from the original
+    provider, and wrapping each method with a fallthrough method that calls ``request_forwarder`` if the
+    original provider raises a ``NotImplementedError``.
+
+    :param provider: the ASF provider
+    :param request_forwarder: callable that forwards the request (e.g., to a backend server)
+    :return: a modified DispatchTable
+    """
+    table = create_dispatch_table(provider)
+
+    for op, fn in table.items():
+        table[op] = _wrap_with_fallthrough(fn, request_forwarder)
+
+    return table
+
+
+def _wrap_with_fallthrough(
+    handler: ServiceRequestHandler, request_forwarder: RequestForwarder
+) -> ServiceRequestHandler:
+    def _call(context, req) -> ServiceResponse:
+        try:
+            # handler will typically be an ASF provider method, and in case it hasn't been
+            # implemented, we try to fall back to forwarding the request to the backend
+            return handler(context, req)
+        except NotImplementedError:
+            return request_forwarder(context)
+
+    return _call
+
+
+def ExternalProcessFallbackDispatcher(provider: object, forward_url_getter: Callable[[], str]):
+    return ForwardingFallbackDispatcher(provider, get_request_forwarder_http(forward_url_getter))
+
+
+def get_request_forwarder_http(forward_url_getter: Callable[[], str]) -> RequestForwarder:
```

Comment:
So, the main reason for the factory is to keep the flexibility of handling changing target ports over time. This may be required in the persistence reset case, where we need to restart an external process to reload the state from disk.

Technically, we could probably make it work with a static port (restarting the process on the same port), but that may introduce some limitations, as we're currently allocating random ports for the external servers. 

Btw, this also relates to the [comment about making the server part of the provider](https://github.com/localstack/localstack/pull/5796#discussion_r843049547) - we can revisit once we have more clarity around the restarting / persistence integration. 👍